File: java/src/main/java/shop/pricing/Money.java
package shop.pricing;

/** Money in minor units. A record, so equality is by value. */
public record Money(int cents) {
    public static final Money ZERO = new Money(0);

    /* addExact and multiplyExact, not + and *: Java's int arithmetic wraps
       silently, so an overflowing total becomes a large negative amount and
       nothing anywhere reports a problem. On money that is the worst
       possible failure mode, and the test for it is one line. */
    public Money plus(Money other) {
        return new Money(Math.addExact(cents, other.cents));
    }

    public Money times(int quantity) {
        return new Money(Math.multiplyExact(cents, quantity));
    }

    @Override
    public String toString() {
        return "£%d.%02d".formatted(cents / 100, Math.abs(cents % 100));
    }
}

File: dotnet/Shop/Pricing/Money.cs
namespace Shop.Pricing;

/// <summary>
/// Money in minor units. A readonly record struct so equality is by value
/// and a test can assert on it directly.
/// </summary>
public readonly record struct Money(int Cents)
{
    public static Money Zero => new(0);

    public static Money operator +(Money left, Money right) => new(left.Cents + right.Cents);

    public static Money operator *(Money money, int quantity) => new(money.Cents * quantity);

    public override string ToString() => $"£{Cents / 100m:0.00}";
}
